types/canonical.go

```diff
@@ -23,32 +23,32 @@ type CanonicalPartSetHeader struct {
 }
 
 type CanonicalProposal struct {
+	Version          uint64                 `json:"version" binary:"fixed64"`
+	Height           int64                  `json:"height" binary:"fixed64"`
+	Round            int64                  `json:"round" binary:"fixed64"`
+	POLRound         int64                  `json:"pol_round" binary:"fixed64"`
+	Timestamp        time.Time              `json:"timestamp"`
 	ChainID          string                 `json:"@chain_id"`
-	Type             string                 `json:"@type"`
```

Comment:
I guess one problem with removing `type` is that we have to be careful that eg. the encoding of a CanonicalProposal could not be mistaken for that of a CanonicalVote. 

Are we certain that `marhsalBinary(x) != marshalBinary(y)` for `var x CanonicalVote` and `var y CanonicalProposal` ?

Reply:
You are right. They would actually be the same if for instance the first 3 fields are equal (because of the same order and nothing else to differentiate between the two). I'll add a test-case and add back the type.

Reply:
Thats much better. Thanks for the thorough review and the great suggestions ❤️  
This change will touch quite a lot of files though. I'll rebase (and resolve conflicts) and this should be ready for another review.